massa-bootstrap/src/server.rs

```diff
@@ -0,0 +1,455 @@
+        */
+        loop {
+            massa_trace!("bootstrap.lib.run.select", {});
+            tokio::select! {
+                // managed commands
+                _ = self.manager_rx.recv() => {
+                    massa_trace!("bootstrap.lib.run.select.manager", {});
+                    break
+                },
+
+                // cache cleanup timeout
+                _ = &mut cache_timer, if bootstrap_data.is_some() => {
+                    massa_trace!("bootstrap.lib.run.cache_unload", {});
+                    bootstrap_data = None;
+                }
+
+                // bootstrap session finished
+                Some(_) = bootstrap_sessions.next() => {
+                    massa_trace!("bootstrap.session.finished", {"active_count": bootstrap_sessions.len()});
+                }
+
+                // listener
+                Ok((dplx, remote_addr)) = listener.accept() => if bootstrap_sessions.len() < self.bootstrap_settings.max_simultaneous_bootstraps as usize {
+                    massa_trace!("bootstrap.lib.run.select.accept", {"remote_addr": remote_addr});
+                    let now = Instant::now();
+
+                    // clear IP history if necessary
+                    if self.ip_hist_map.len() > self.bootstrap_settings.ip_list_max_size {
+                        self.ip_hist_map.retain(|_k, v| now.duration_since(*v) <= per_ip_min_interval);
+                        if self.ip_hist_map.len() > self.bootstrap_settings.ip_list_max_size {
+                            // too many IPs are spamming us: clear cache
+                            warn!("high bootstrap load: at least {} different IPs attempted bootstrap in the last {}ms", self.ip_hist_map.len(), self.bootstrap_settings.per_ip_min_interval);
+                            self.ip_hist_map.clear();
+                        }
+                    }
+
+                    // check IP's bootstrap attempt history
+                    match self.ip_hist_map.entry(remote_addr.ip()) {
+                        hash_map::Entry::Occupied(mut occ) => {
+                            if now.duration_since(*occ.get()) <= per_ip_min_interval {
+                                let mut server = BootstrapServerBinder::new(dplx, self.private_key);
+                                let _ = match tokio::time::timeout(self.bootstrap_settings.write_error_timeout.into(), server.send(BootstrapMessageServer::BootstrapError {
+                                    error:
+                                    format!("Your last bootstrap on this server was {:#?} ago and you have to wait {:#?} before retrying.", occ.get().elapsed(), per_ip_min_interval.saturating_sub(occ.get().elapsed()))
+                                })).await {
+                                    Err(_) => Err(std::io::Error::new(std::io::ErrorKind::TimedOut, "bootstrap error no available slots send timed out").into()),
+                                    Ok(Err(e)) => Err(e),
```

Comment:
Shouldn't this ban the node?

Reply:
No we don't ban nodes in bootstrap process. The node isn't a peer at this time.